libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/fakeResolver.kt

```diff
@@ -266,14 +277,30 @@ class DefaultFakeResolver(val types: List<CompiledNamedType>) : FakeResolver {
 
   override fun resolveTypename(context: FakeResolverContext): String {
     val rawType = context.mergedField.type.rawType()
-    val name = rawType.name
-    val index = impl.getOrElse(name) { 0 }
-
-    impl[name] = index + 1
 
     // XXX: Cache this computation
-    val possibleTypes = possibleTypes(types, rawType)
-    return possibleTypes[index % possibleTypes.size].name
+    val possibleTypes = possibleTypes(allTypes, rawType)
+    val index = context.id.hashCode().mod(possibleTypes.size)
+    return possibleTypes[index].name
+  }
+
+  override fun stableIdForObject(obj: Map<String, Any?>, mergedField: CompiledField): String? {
+    val keyFields = mergedField.type.rawType().keyFields()
+
+    if (keyFields.isNotEmpty()) {
+      return buildString {
+        append(obj["__typename"].toString())
+        keyFields.forEach {
+          append(obj[it].toString())
+        }
+      }
+    }
+
+    if (obj.containsKey("__stableId")) {
```

Comment:
Should this test be done first? If for some reason one wants to override the stable id even when there's a typepolicy (not sure why though).

Reply:
Good call, I'll change this 👍